File: Models/BiomaSurfaceSWATSoilSWATC/src/simplace/BiomaSurfaceSWATSoilSWATC/SoilTemperatureSWAT.java
package net.simplace.usermodules.amei.Bioma;
import  java.util.*;
import net.simplace.sim.model.FWSimComponent;
import net.simplace.sim.util.FWSimVarMap;
import net.simplace.sim.util.FWSimVariable;
import net.simplace.sim.util.FWSimVariable.CONTENT_TYPE;
import net.simplace.sim.util.FWSimVariable.DATA_TYPE;
import org.jdom2.Element;


public class SoilTemperatureSWAT extends FWSimComponent
{
    private FWSimVariable<Double[]> VolumetricWaterContent;
    private FWSimVariable<Double> SurfaceSoilTemperature;
    private FWSimVariable<Double[]> LayerThickness;
    private FWSimVariable<Double> LagCoefficient;
    private FWSimVariable<Double[]> SoilTemperatureByLayers;
    private FWSimVariable<Double> AirTemperatureAnnualAverage;
    private FWSimVariable<Double[]> BulkDensity;
    private FWSimVariable<Double> SoilProfileDepth;

    public SoilTemperatureSWAT(String aName, HashMap<String, FWSimVariable<?>> aFieldMap, HashMap<String, String> aInputMap, Element aSimComponentElement, FWSimVarMap aVarMap, int aOrderNumber)
    {
        super(aName, aFieldMap, aInputMap, aSimComponentElement, aVarMap, aOrderNumber);
    }

    public SoilTemperatureSWAT(){
        super();
    }

    @Override
    public HashMap<String, FWSimVariable<?>> createVariables()
    {
        addVariable(FWSimVariable.createSimVariable("VolumetricWaterContent", "Volumetric soil water content", DATA_TYPE.DOUBLEARRAY, CONTENT_TYPE.input,"m3 m-3", 0, 0.8, null, this));
        addVariable(FWSimVariable.createSimVariable("SurfaceSoilTemperature", "Average surface soil temperature", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"degC", -60d, 60d, 25d, this));
        addVariable(FWSimVariable.createSimVariable("LayerThickness", "Soil layer thickness", DATA_TYPE.DOUBLEARRAY, CONTENT_TYPE.constant,"m", 0.005, 3d, null, this));
        addVariable(FWSimVariable.createSimVariable("LagCoefficient", "Lag coefficient that controls the influence of the previous day's temperature on the current day's temperature", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"dimensionless", 0, 1, 0.8, this));
        addVariable(FWSimVariable.createSimVariable("SoilTemperatureByLayers", "Soil temperature of each layer", DATA_TYPE.DOUBLEARRAY, CONTENT_TYPE.state,"degC", -60, 60, null, this));
        addVariable(FWSimVariable.createSimVariable("AirTemperatureAnnualAverage", "Annual average air temperature", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"degC", -40d, 50d, 15d, this));
        addVariable(FWSimVariable.createSimVariable("BulkDensity", "Bulk density", DATA_TYPE.DOUBLEARRAY, CONTENT_TYPE.constant,"t m-3", 0.9, 1.8, null, this));
        addVariable(FWSimVariable.createSimVariable("SoilProfileDepth", "Soil profile depth", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"m", 0, 50d, 3d, this));

        return iFieldMap;
    }
    @Override
    protected void init()
    {
        Double [] t_VolumetricWaterContent = VolumetricWaterContent.getValue();
        Double [] t_LayerThickness = LayerThickness.getValue();
        double t_LagCoefficient = LagCoefficient.getValue();
        double t_AirTemperatureAnnualAverage = AirTemperatureAnnualAverage.getValue();
        Double [] t_BulkDensity = BulkDensity.getValue();
        double t_SoilProfileDepth = SoilProfileDepth.getValue();
        Double [] t_SoilTemperatureByLayers = new Double[t_LayerThickness.length];
        Integer i;
        Arrays.fill(t_SoilTemperatureByLayers, 0.0d);
        for (i=0 ; i!=t_LayerThickness.length ; i+=1)
        {
            t_SoilTemperatureByLayers[i] = (double)(15);
        }
        SoilTemperatureByLayers.setValue(t_SoilTemperatureByLayers, this);
    }
    @Override
    protected void process()
    {
        Double [] t_VolumetricWaterContent = VolumetricWaterContent.getValue();
        double t_SurfaceSoilTemperature = SurfaceSoilTemperature.getValue();
        Double [] t_LayerThickness = LayerThickness.getValue();
        double t_LagCoefficient = LagCoefficient.getValue();
        Double [] t_SoilTemperatureByLayers = SoilTemperatureByLayers.getValue();
        double t_AirTemperatureAnnualAverage = AirTemperatureAnnualAverage.getValue();
        Double [] t_BulkDensity = BulkDensity.getValue();
        double t_SoilProfileDepth = SoilProfileDepth.getValue();
        Integer i;
        double _SoilProfileDepthmm;
        double _TotalWaterContentmm;
        double _MaximumDumpingDepth;
        double _DumpingDepth;
        double _ScalingFactor;
        double _DepthBottom;
        double _RatioCenter;
        double _DepthFactor;
        double _DepthCenterLayer;
        _SoilProfileDepthmm = t_SoilProfileDepth * 1000;
        _TotalWaterContentmm = (double)(0);
        for (i=0 ; i!=t_LayerThickness.length ; i+=1)
        {
            _TotalWaterContentmm = _TotalWaterContentmm + (t_VolumetricWaterContent[i] * t_LayerThickness[i]);
        }
        _TotalWaterContentmm = _TotalWaterContentmm * 1000;
        _MaximumDumpingDepth = (double)(0);
        _DumpingDepth = (double)(0);
        _ScalingFactor = (double)(0);
        _DepthBottom = (double)(0);
        _RatioCenter = (double)(0);
        _DepthFactor = (double)(0);
        _DepthCenterLayer = t_LayerThickness[0] * 1000 / 2;
        _MaximumDumpingDepth = 1000 + (2500 * t_BulkDensity[0] / (t_BulkDensity[0] + (686 * Math.exp(-5.63d * t_BulkDensity[0]))));
        _ScalingFactor = _TotalWaterContentmm / ((0.356d - (0.144d * t_BulkDensity[0])) * _SoilProfileDepthmm);
        _DumpingDepth = _MaximumDumpingDepth * Math.exp(Math.log(500 / _MaximumDumpingDepth) * Math.pow((1 - _ScalingFactor) / (1 + _ScalingFactor), 2));
        _RatioCenter = _DepthCenterLayer / _DumpingDepth;
        _DepthFactor = _RatioCenter / (_RatioCenter + Math.exp(-0.867d - (2.078d * _RatioCenter)));
        t_SoilTemperatureByLayers[0] = t_LagCoefficient * t_SoilTemperatureByLayers[0] + ((1 - t_LagCoefficient) * (_DepthFactor * (t_AirTemperatureAnnualAverage - t_SurfaceSoilTemperature) + t_SurfaceSoilTemperature));
        for (i=1 ; i!=t_LayerThickness.length ; i+=1)
        {
            _DepthBottom = _DepthBottom + (t_LayerThickness[(i - 1)] * 1000);
            _DepthCenterLayer = _DepthBottom + (t_LayerThickness[i] * 1000 / 2);
            _MaximumDumpingDepth = 1000 + (2500 * t_BulkDensity[i] / (t_BulkDensity[i] + (686 * Math.exp(-5.63d * t_BulkDensity[i]))));
            _ScalingFactor = _TotalWaterContentmm / ((0.356d - (0.144d * t_BulkDensity[i])) * _SoilProfileDepthmm);
            _DumpingDepth = _MaximumDumpingDepth * Math.exp(Math.log(500 / _MaximumDumpingDepth) * Math.pow((1 - _ScalingFactor) / (1 + _ScalingFactor), 2));
            _RatioCenter = _DepthCenterLayer / _DumpingDepth;
            _DepthFactor = _RatioCenter / (_RatioCenter + Math.exp(-0.867d - (2.078d * _RatioCenter)));
            t_SoilTemperatureByLayers[i] = t_LagCoefficient * t_SoilTemperatureByLayers[i] + ((1 - t_LagCoefficient) * (_DepthFactor * (t_AirTemperatureAnnualAverage - t_SurfaceSoilTemperature) + t_SurfaceSoilTemperature));
        }
        SoilTemperatureByLayers.setValue(t_SoilTemperatureByLayers, this);
    }

    public HashMap<String, FWSimVariable<?>> fillTestVariables(int aParamIndex, TEST_STATE aDefineOrCheck)
    {
        return iFieldMap;
    }

    @Override
    protected FWSimComponent clone(FWSimVarMap aVarMap)
    {
        return new SoilTemperatureSWAT(iName, iFieldMap, iInputMap, iSimComponentElement, aVarMap, iOrderNumber);
    }
}

File: Models/BiomaSurfaceSWATSoilSWATC/src/apsim/BiomaSurfaceSWATSoilSWATC/SoilTemperatureSWAT.cs
using System;
using System.Collections.Generic;
using System.Linq;    
using Models.Core;   
namespace Models.Crop2ML;

/// <summary>
///- Name: SoilTemperatureSWAT -Version: 001, -Time step: 1
///- Description:
///            * Title: SoilTemperatureSWAT model
///            * Authors: simone.bregaglio
///            * Reference: http://bioma.jrc.ec.europa.eu/ontology/JRC_MARS_biophysical_domain.owl
///            * Institution: University Of Milan
///            * ExtendedDescription: Strategy for the calculation of soil temperature with SWAT method. Reference: Neitsch,S.L., Arnold, J.G., Kiniry, J.R., Williams, J.R., King, K.W. Soil and Water Assessment Tool. Theoretical documentation. Version 2000. http://swatmodel.tamu.edu/media/1290/swat2000theory.pdf
///            * ShortDescription: Strategy for the calculation of soil temperature with SWAT method
///- inputs:
///            * name: VolumetricWaterContent
///                          ** description : Volumetric soil water content
///                          ** inputtype : variable
///                          ** variablecategory : exogenous
///                          ** datatype : DOUBLEARRAY
///                          ** len : 
///                          ** max : 0.8
///                          ** min : 0
///                          ** default : 0.25
///                          ** unit : m3 m-3
///            * name: SurfaceSoilTemperature
///                          ** description : Average surface soil temperature
///                          ** inputtype : variable
///                          ** variablecategory : auxiliary
///                          ** datatype : DOUBLE
///                          ** max : 60
///                          ** min : -60
///                          ** default : 25
///                          ** unit : degC
///            * name: LayerThickness
///                          ** description : Soil layer thickness
///                          ** inputtype : parameter
///                          ** parametercategory : constant
///                          ** datatype : DOUBLEARRAY
///                          ** len : 
///                          ** max : 3
///                          ** min : 0.005
///                          ** default : 0.05
///                          ** unit : m
///            * name: LagCoefficient
///                          ** description : Lag coefficient that controls the influence of the previous day's temperature on the current day's temperature
///                          ** inputtype : parameter
///                          ** parametercategory : constant
///                          ** datatype : DOUBLE
///                          ** max : 1
///                          ** min : 0
///                          ** default : 0.8
///                          ** unit : dimensionless
///            * name: SoilTemperatureByLayers
///                          ** description : Soil temperature of each layer
///                          ** inputtype : variable
///                          ** variablecategory : state
///                          ** datatype : DOUBLEARRAY
///                          ** len : 
///                          ** max : 60
///                          ** min : -60
///                          ** default : 15
///                          ** unit : degC
///            * name: AirTemperatureAnnualAverage
///                          ** description : Annual average air temperature
///                          ** inputtype : parameter
///                          ** parametercategory : constant
///                          ** datatype : DOUBLE
///                          ** max : 50
///                          ** min : -40
///                          ** default : 15
///                          ** unit : degC
///            * name: BulkDensity
///                          ** description : Bulk density
///                          ** inputtype : parameter
///                          ** parametercategory : constant
///                          ** datatype : DOUBLEARRAY
///                          ** len : 
///                          ** max : 1.8
///                          ** min : 0.9
///                          ** default : 1.3
///                          ** unit : t m-3
///            * name: SoilProfileDepth
///                          ** description : Soil profile depth
///                          ** inputtype : parameter
///                          ** parametercategory : constant
///                          ** datatype : DOUBLE
///                          ** max : 50
///                          ** min : 0
///                          ** default : 3
///                          ** unit : m
///- outputs:
///            * name: SoilTemperatureByLayers
///                          ** description : Soil temperature of each layer
///                          ** datatype : DOUBLEARRAY
///                          ** variablecategory : state
///                          ** len : 
///                          ** max : 60
///                          ** min : -60
///                          ** unit : degC
/// </summary>
public class SoilTemperatureSWAT
{

    /// <summary>
    /// initialization of the SoilTemperatureSWAT component
    /// </summary>
    public void Init(SurfaceSWATSoilSWATCState s, SurfaceSWATSoilSWATCState s1, SurfaceSWATSoilSWATCRate r, SurfaceSWATSoilSWATCAuxiliary a, SurfaceSWATSoilSWATCExogenous ex)
    {
        double[] VolumetricWaterContent = ex.VolumetricWaterContent;
        double[] SoilTemperatureByLayers ;
        int i;
        SoilTemperatureByLayers = new double[LayerThickness.Length];
        for (i=0 ; i!=LayerThickness.Length ; i+=1)
        {
            SoilTemperatureByLayers[i] = (double)(15);
        }
        s.SoilTemperatureByLayers= SoilTemperatureByLayers;
    }

    private double[] _LayerThickness;
    /// <summary>
    /// Gets and sets the Soil layer thickness
    /// </summary>
    [Description("Soil layer thickness")] 
    [Units("m")] 
    //[Crop2ML(datatype="DOUBLEARRAY", min=0.005, max=3, default=0.05, parametercategory=constant, inputtype="parameter")] 
    public double[] LayerThickness
    {
        get { return this._LayerThickness; }
        set { this._LayerThickness= value; } 
    }

    private double _LagCoefficient;
    /// <summary>
    /// Gets and sets the Lag coefficient that controls the influence of the previous day's temperature on the current day's temperature
    /// </summary>
    [Description("Lag coefficient that controls the influence of the previous day's temperature on the current day's temperature")] 
    [Units("dimensionless")] 
    //[Crop2ML(datatype="DOUBLE", min=0, max=1, default=0.8, parametercategory=constant, inputtype="parameter")] 
    public double LagCoefficient
    {
        get { return this._LagCoefficient; }
        set { this._LagCoefficient= value; } 
    }

    private double _AirTemperatureAnnualAverage;
    /// <summary>
    /// Gets and sets the Annual average air temperature
    /// </summary>
    [Description("Annual average air temperature")] 
    [Units("degC")] 
    //[Crop2ML(datatype="DOUBLE", min=-40, max=50, default=15, parametercategory=constant, inputtype="parameter")] 
    public double AirTemperatureAnnualAverage
    {
        get { return this._AirTemperatureAnnualAverage; }
        set { this._AirTemperatureAnnualAverage= value; } 
    }

    private double[] _BulkDensity;
    /// <summary>
    /// Gets and sets the Bulk density
    /// </summary>
    [Description("Bulk density")] 
    [Units("t m-3")] 
    //[Crop2ML(datatype="DOUBLEARRAY", min=0.9, max=1.8, default=1.3, parametercategory=constant, inputtype="parameter")] 
    public double[] BulkDensity
    {
        get { return this._BulkDensity; }
        set { this._BulkDensity= value; } 
    }

    private double _SoilProfileDepth;
    /// <summary>
    /// Gets and sets the Soil profile depth
    /// </summary>
    [Description("Soil profile depth")] 
    [Units("m")] 
    //[Crop2ML(datatype="DOUBLE", min=0, max=50, default=3, parametercategory=constant, inputtype="parameter")] 
    public double SoilProfileDepth
    {
        get { return this._SoilProfileDepth; }
        set { this._SoilProfileDepth= value; } 
    }

    
    /// <summary>
    /// Constructor of the SoilTemperatureSWAT component")
    /// </summary>  
    public SoilTemperatureSWAT() { }
    
    /// <summary>
    /// Algorithm of the SoilTemperatureSWAT component
    /// </summary>
    public void  CalculateModel(SurfaceSWATSoilSWATCState s, SurfaceSWATSoilSWATCState s1, SurfaceSWATSoilSWATCRate r, SurfaceSWATSoilSWATCAuxiliary a, SurfaceSWATSoilSWATCExogenous ex)
    {
        double[] VolumetricWaterContent = ex.VolumetricWaterContent;
        double SurfaceSoilTemperature = a.SurfaceSoilTemperature;
        double[] SoilTemperatureByLayers = s.SoilTemperatureByLayers;
        int i;
        double _SoilProfileDepthmm;
        double _TotalWaterContentmm;
        double _MaximumDumpingDepth;
        double _DumpingDepth;
        double _ScalingFactor;
        double _DepthBottom;
        double _RatioCenter;
        double _DepthFactor;
        double _DepthCenterLayer;
        _SoilProfileDepthmm = SoilProfileDepth * 1000;
        _TotalWaterContentmm = (double)(0);
        for (i=0 ; i!=LayerThickness.Length ; i+=1)
        {
            _TotalWaterContentmm = _TotalWaterContentmm + (VolumetricWaterContent[i] * LayerThickness[i]);
        }
        _TotalWaterContentmm = _TotalWaterContentmm * 1000;
        _MaximumDumpingDepth = (double)(0);
        _DumpingDepth = (double)(0);
        _ScalingFactor = (double)(0);
        _DepthBottom = (double)(0);
        _RatioCenter = (double)(0);
        _DepthFactor = (double)(0);
        _DepthCenterLayer = LayerThickness[0] * 1000 / 2;
        _MaximumDumpingDepth = 1000 + (2500 * BulkDensity[0] / (BulkDensity[0] + (686 * Math.Exp(-5.63 * BulkDensity[0]))));
        _ScalingFactor = _TotalWaterContentmm / ((0.356 - (0.144 * BulkDensity[0])) * _SoilProfileDepthmm);
        _DumpingDepth = _MaximumDumpingDepth * Math.Exp(Math.Log(500 / _MaximumDumpingDepth) * Math.Pow((1 - _ScalingFactor) / (1 + _ScalingFactor), 2));
        _RatioCenter = _DepthCenterLayer / _DumpingDepth;
        _DepthFactor = _RatioCenter / (_RatioCenter + Math.Exp(-0.867 - (2.078 * _RatioCenter)));
        SoilTemperatureByLayers[0] = LagCoefficient * SoilTemperatureByLayers[0] + ((1 - LagCoefficient) * (_DepthFactor * (AirTemperatureAnnualAverage - SurfaceSoilTemperature) + SurfaceSoilTemperature));
        for (i=1 ; i!=LayerThickness.Length ; i+=1)
        {
            _DepthBottom = _DepthBottom + (LayerThickness[(i - 1)] * 1000);
            _DepthCenterLayer = _DepthBottom + (LayerThickness[i] * 1000 / 2);
            _MaximumDumpingDepth = 1000 + (2500 * BulkDensity[i] / (BulkDensity[i] + (686 * Math.Exp(-5.63 * BulkDensity[i]))));
            _ScalingFactor = _TotalWaterContentmm / ((0.356 - (0.144 * BulkDensity[i])) * _SoilProfileDepthmm);
            _DumpingDepth = _MaximumDumpingDepth * Math.Exp(Math.Log(500 / _MaximumDumpingDepth) * Math.Pow((1 - _ScalingFactor) / (1 + _ScalingFactor), 2));
            _RatioCenter = _DepthCenterLayer / _DumpingDepth;
            _DepthFactor = _RatioCenter / (_RatioCenter + Math.Exp(-0.867 - (2.078 * _RatioCenter)));
            SoilTemperatureByLayers[i] = LagCoefficient * SoilTemperatureByLayers[i] + ((1 - LagCoefficient) * (_DepthFactor * (AirTemperatureAnnualAverage - SurfaceSoilTemperature) + SurfaceSoilTemperature));
        }
        s.SoilTemperatureByLayers= SoilTemperatureByLayers;
    }
}